File: BSIGeneralAffairDAL/DALBrand.vb
Imports BSIGeneralAffairBO
Imports GeneralAffairInterface
Imports System.Data
Imports System.Data.SqlClient

Public Class DALBrand
    Implements IBrand

    Public database As DatabaseConfiguration = New DatabaseConfiguration()
    Private dataRead As SqlDataReader
    Public cmd As SqlCommand

    Public Function getAll() As List(Of Brand) Implements IBrand.getAll
        Dim Brands As New List(Of Brand)
        Try
            Dim strSP = "[GeneralAffair].[USP_SelectAllBrand]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                While dataRead.Read
                    Dim Brand As New Brand
                    Brand.BrandID = CInt(dataRead("BrandID"))
                    Brand.BrandName = dataRead("BrandName").ToString()
                    Brand.UpdatedAt = dataRead("UpdatedAt").ToString()
                    Brands.Add(Brand)
                End While
            End If
            dataRead.Close()

            Return Brands
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function create(obj As Brand) As Integer Implements IBrand.create
        Try
            Dim strSP = "[GeneralAffair].[USP_StoreBrand]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure

            cmd.Parameters.AddWithValue("@BrandName", obj.BrandName)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            If result = 1 Then
                Throw New ArgumentException("Brand not created")
            End If
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function update(obj As Brand) As Integer Implements IBrand.update
        Try
            Dim updateBrand = GetById(obj.BrandID)
            If updateBrand Is Nothing Then
                Throw New ArgumentException("Brand not found")
            End If

            Dim strSP = "[GeneralAffair].[USP_UpdateBrand]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@BrandID", obj.BrandID)
            cmd.Parameters.AddWithValue("@BrandName", obj.BrandName)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Private Function GetById(brandID As Integer) As Object
        Try
            cmd = New SqlCommand("SELECT * FROM [GeneralAffair].[Brands] WHERE BrandID = @BrandID", database.conn)
            cmd.Parameters.AddWithValue("@BrandID", brandID)
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                dataRead.Read()
                Dim brand As New Brand
                brand.BrandID = CInt(dataRead("BrandID"))
                brand.BrandName = dataRead("BrandName").ToString()
                brand.UpdatedAt = dataRead("UpdatedAt").ToString()
                Return brand
            Else
                Return Nothing
            End If
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function delete(id As Integer) As Integer Implements IBrand.delete
        Try
            Dim updateBrand = GetById(id)
            If updateBrand Is Nothing Then
                Throw New ArgumentException("Brand not found")
            End If

            Dim strSP = "[GeneralAffair].[USP_DeleteBrand]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@BrandID", id)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function
End Class

File: BSIGeneralAffairDAL-C/DALBrand.cs
using BSIGeneralAffairBO;
using BSIGeneralAffairDAL_C.Interfaces;
using GeneralAffairInterface;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SqlClient;
using System.Text;
using static Dapper.SqlMapper;

namespace BSIGeneralAffairDAL_C
{
    public class DALBrand : IBrandDAL
    {
        public void Insert(Brand entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = "[GeneralAffair].[USP_StoreBrand]";
                var param = new { BrandName = entity.BrandName };
                try
                {
                    int result = conn.Execute(strSql, param, commandType: System.Data.CommandType.StoredProcedure);
                    if (result == 1)
                    {
                        throw new ArgumentException("Insert data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        public void Delete(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[GeneralAffair].[USP_DeleteBrand]";
                var param = new { BrandID = id };
                try
                {
                    int result = conn.Execute(strSql, param);
                    if (result == 1)
                    {
                        throw new ArgumentException("Delete data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        public IEnumerable<Brand> GetAll()
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                //var strSql = @"select * from Categories order by CategoryName";
                var strSql = @"[GeneralAffair].[USP_SelectAllBrand]";

                var results = conn.Query<Brand>(strSql, commandType: System.Data.CommandType.StoredProcedure);
                return results;
            }
        }

        public Brand GetById(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[Brands] Where [BrandID] = @BrandID ";
                var param = new { BrandID = id };
                var result = conn.QuerySingleOrDefault<Brand>(strSql, param);
                return result;
            }
        }

        public IEnumerable<Brand> GetByName(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[Brands] Where [BrandName] = @BrandName ";
                var param = new { BrandName = name };
                var result = conn.Query<Brand>(strSql, param);
                return result;
            }
        }

        public void Update(Brand entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                try
                {
                    var strSql = @"[GeneralAffair].[USP_UpdateBrand]";
                    var param = new { BrandID = entity.BrandID , BrandName = entity.BrandName };
                    int result = conn.Execute(strSql, param);

                    //jika result = -1, berarti update data gagal
                    if (result == 1)
                    {
                        throw new Exception("Update data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new Exception($"{sqlEx.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new Exception(ex.Message);
                }
            }
        }

        private string GetConnectionString()
        {
            return Helper.GetConnectionString();
            //return @"Data Source=ACTUAL;Initial Catalog=LatihanDb;Integrated Security=True;TrustServerCertificate=True";
            //return ConfigurationManager.ConnectionStrings["MyDbConnectionString"].ConnectionString;
        }

        public int GetCountBrands(string name) 
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString())) 
            {
                var strSql = @"SELECT COUNT(*) FROM [GeneralAffair].[Brands] Where [BrandName] like @BrandName ";
                var param = new { BrandName = $"%{name}%" };
                var result = Convert.ToInt32(conn.ExecuteScalar(strSql, param));
                return result;
            }
        }

        public IEnumerable<Brand> GetWithPaging(int pageNumber, int pageSize, string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString())) 
            {
                var strSql = @"SELECT * FROM [GeneralAffair].[Brands] 
                              Where [BrandName] like @BrandName 
                              order by BrandName OFFSET @Offset ROWS FETCH NEXT @PageSize ROWS ONLY";
                var param = new { BrandName = $"%{name}%", Offset = (pageNumber - 1) * pageSize, PageSize = pageSize };
                var results = conn.Query<Brand>(strSql, param);
                return results;
            }
        }
    }
}
